Code/GraphMol/SubstructLibrary/SubstructLibrary.h

```diff
@@ -753,6 +806,17 @@ class RDKIT_SUBSTRUCTLIBRARY_EXPORT SubstructLibrary {
   const std::vector<unsigned int> &getSearchOrder() const {
     return searchOrder;
   }
+  const std::string &getKey(unsigned int idx) const {
+    // expects implementation to throw IndexError if out of range
+    PRECONDITION(keyholder.get() != nullptr, "SubstructLibrary Has no keys");
+    return keyholder->getKey(idx);
+  }
+  std::vector<std::string> getKeys(const std::vector<unsigned int> &indices) const {
```

Comment:
This is inconsistent with `getMolecules()`, which returns a reference to the internal MolHolder.
I think it would be better to use that pattern here, but to return a pointer to the KeyHolder since it's not guaranteed to actually be there.

Reply:
I think this is going to be used enough I'd like to have an API point on the substruct library.  I can see getKeys is probably an inconsistent usage.

What do you think of an overload as a convenience?

```
keyholder keys = lib.getKeys();
ids = lib.getKeys(matches);
```

Reply:
The alternative is ok but kind of ugly:

```
lib.getKeys().getKeys(matches);
```

Reply:
I guess I don't see how the usage of this is any different from the usage of the MolHolder...
How would you explain that.

Reply:
The Molholder doesn't return multiple molecules mostly because of potential size limitations.  I'm happy removing the overload for now, it's easy to put back.

Reply:
> The alternative is ok but kind of ugly:
> 
> ```
> lib.getKeys().getKeys(matches);
> ```

I think that we made a mistake with the original API for `SubstructLibrary`: it was a mistake to have a method called `getMolecules()` which returns a `MolHolder` object... that should have been called `getMolHolder()`. But that's where we are.

For this new method I would prefer `lib.getKeyHolder().getKeys()` over `lib.getKeys().getKeys()`, but I find both of them to be considerably less ugly than having `getKeys()` overloads which return things that are semantically totally different (one returns keys, the other the object which contains the keys). But if you have a strong preference, go for it.